File: app/src/main/java/tn/esprit/wayfinder/viewmodels/OnboardingViewModel.kt
package tn.esprit.wayfinder.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import tn.esprit.wayfinder.models.AnswerRequest
import tn.esprit.wayfinder.models.OnboardingQuestion
import tn.esprit.wayfinder.models.Progress
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import retrofit2.HttpException
import tn.esprit.wayfinder.presentation.auth.OnboardingRepository
import java.util.concurrent.atomic.AtomicInteger

sealed class OnboardingUiState {
    object Idle : OnboardingUiState()
    object Loading : OnboardingUiState()
    data class QuestionLoaded(
        val question: OnboardingQuestion,
        val progress: Progress
    ) : OnboardingUiState()
    data class Completed(val message: String) : OnboardingUiState()
    data class Error(val message: String) : OnboardingUiState()
}

data class OnboardingSyncStatus(
    val questionsAnswered: Int,
    val canResume: Boolean,
    val lastUpdated: Long = System.currentTimeMillis()
)

class OnboardingViewModel(private val onboardingRepository: OnboardingRepository) : ViewModel() {

    private val _uiState = MutableStateFlow<OnboardingUiState>(OnboardingUiState.Idle)
    val uiState: StateFlow<OnboardingUiState> = _uiState.asStateFlow()

    private val _syncStatus = MutableStateFlow<OnboardingSyncStatus?>(null)
    val syncStatus: StateFlow<OnboardingSyncStatus?> = _syncStatus.asStateFlow()

    private var currentSessionId: String? = null
    private val operationCounter = AtomicInteger(0)
    @Volatile private var latestOperationId = 0

    private fun beginOperation(): Int {
        val id = operationCounter.incrementAndGet()
        latestOperationId = id
        return id
    }

    private fun isStaleOperation(operationId: Int): Boolean {
        return operationId != latestOperationId
    }

    fun startOnboarding() {
        val operationId = beginOperation()
        viewModelScope.launch {
            try {
                _uiState.value = OnboardingUiState.Loading
                val response = onboardingRepository.startOnboarding()
                if (isStaleOperation(operationId)) return@launch
                currentSessionId = response.sessionId
                if (response.completed) {
                    _uiState.value = OnboardingUiState.Completed(
                        message = response.message ?: "Onboarding completed!"
                    )
                } else {
                    _uiState.value = OnboardingUiState.QuestionLoaded(
                        question = requireNotNull(response.question),
                        progress = requireNotNull(response.progress)
                    )
                }
                verifyProgress()
            } catch (e: Exception) {
                if (isStaleOperation(operationId)) return@launch
                // Special handling: backend returns 400 when onboarding is already completed
                if (e is HttpException) {
                    val errorBody = e.response()?.errorBody()?.string()
                    val message = try {
                        errorBody?.let {
                            val element = Json.parseToJsonElement(it)
                            element.jsonObject["message"]?.jsonPrimitive?.content
                        }
                    } catch (_: Exception) {
                        null
                    }

                    if (message == "Onboarding already completed") {
                        _uiState.value = OnboardingUiState.Completed(
                            message = "Onboarding already completed. Redirecting to home..."
                        )
                        return@launch
                    }
                }

                _uiState.value = OnboardingUiState.Error(
                    e.message ?: "Failed to start onboarding"
                )
            }
        }
    }

    fun submitAnswer(questionId: String, answer: Any) {
        val operationId = beginOperation()
        viewModelScope.launch {
            try {
                _uiState.value = OnboardingUiState.Loading
                val sessionId = currentSessionId ?: return@launch

                val serializedAnswer: JsonElement = when (answer) {
                    is String -> JsonPrimitive(answer)
                    is List<*> -> {
                        val strings = answer.filterIsInstance<String>()
                        JsonArray(strings.map { JsonPrimitive(it) })
                    }
                    else -> JsonPrimitive(answer.toString())
                }

                val response = onboardingRepository.submitAnswer(
                    AnswerRequest(sessionId, questionId, serializedAnswer)
                )

                if (isStaleOperation(operationId)) return@launch

                if (response.completed) {
                    _uiState.value = OnboardingUiState.Completed(
                        message = response.message ?: "Onboarding completed!"
                    )
                } else {
                    currentSessionId = response.sessionId
                    _uiState.value = OnboardingUiState.QuestionLoaded(
                        question = requireNotNull(response.question),
                        progress = requireNotNull(response.progress)
                    )
                }
                verifyProgress()
            } catch (e: Exception) {
                if (isStaleOperation(operationId)) return@launch
                _uiState.value = OnboardingUiState.Error(e.message ?: "Failed to submit answer")
            }
        }
    }

    fun skipOnboarding() {
        val operationId = beginOperation()
        viewModelScope.launch {
            try {
                _uiState.value = OnboardingUiState.Loading
                val response = onboardingRepository.skipOnboarding()
                if (isStaleOperation(operationId)) return@launch
                _uiState.value = OnboardingUiState.Completed(
                    message = response.message ?: "Onboarding skipped!"
                )
            } catch (e: Exception) {
                if (isStaleOperation(operationId)) return@launch
                _uiState.value = OnboardingUiState.Error(e.message ?: "Failed to skip onboarding")
            }
        }
    }

    fun resetOnboarding() {
        val operationId = beginOperation()
        viewModelScope.launch {
            try {
                _uiState.value = OnboardingUiState.Loading
                val response = onboardingRepository.resetOnboarding()
                if (isStaleOperation(operationId)) return@launch
                currentSessionId = response.sessionId
                if (response.completed) {
                    _uiState.value = OnboardingUiState.Completed(
                        message = response.message ?: "Onboarding completed!"
                    )
                } else {
                    _uiState.value = OnboardingUiState.QuestionLoaded(
                        question = requireNotNull(response.question),
                        progress = requireNotNull(response.progress)
                    )
                }
                verifyProgress()
            } catch (e: Exception) {
                if (isStaleOperation(operationId)) return@launch
                _uiState.value = OnboardingUiState.Error(e.message ?: "Failed to reset onboarding")
            }
        }
    }

    fun verifyProgress() {
        viewModelScope.launch {
            try {
                val status = onboardingRepository.getStatus()
                currentSessionId = status.sessionId ?: currentSessionId
                _syncStatus.value = OnboardingSyncStatus(
                    questionsAnswered = status.progress.questionsAnswered,
                    canResume = status.canResume
                )
                if (status.onboardingCompleted) {
                    _uiState.value = OnboardingUiState.Completed(
                        message = "Onboarding already completed. Redirecting to home..."
                    )
                }
            } catch (_: Exception) {
                // Ignore sync errors – UI can retry on demand
            }
        }
    }
}

File: WayFinder/ViewModels/OnboardingViewModel.swift
import Foundation
import SwiftUI

enum OnboardingUiState {
    case idle
    case loading
    case questionLoaded(question: OnboardingQuestion, progress: Progress)
    case completed(message: String)
    case error(message: String)
}

@MainActor
class OnboardingViewModel: ObservableObject {
    @Published var uiState: OnboardingUiState = .idle
    @Published var syncStatus: OnboardingSyncStatus?
    
    private var currentSessionId: String?
    private var operationCounter = AtomicInteger()
    private var latestOperationId = 0
    
    struct OnboardingSyncStatus {
        let questionsAnswered: Int
        let canResume: Bool
        let lastUpdated: Date
    }
    
    // MARK: - Start Onboarding
    
    func startOnboarding() async {
        let operationId = beginOperation()
        do {
            // Check status first to avoid "already completed" error
            print("🔄 [OnboardingViewModel] Checking onboarding status before starting...")
            let status = try await OnboardingService.shared.getStatus()
            if status.onboardingCompleted {
                print("⚠️ [OnboardingViewModel] Onboarding already completed! Resetting first...")
                await resetOnboarding()
                return
            }
            
            print("🔄 [OnboardingViewModel] Starting onboarding...")
            uiState = .loading
            let response = try await OnboardingService.shared.startOnboarding()
            
            guard !isStaleOperation(operationId) else { return }
            
            currentSessionId = response.sessionId
            
            if response.completed {
                uiState = .completed(message: response.message ?? "Onboarding completed!")
            } else if let question = response.question, let progress = response.progress {
                uiState = .questionLoaded(question: question, progress: progress)
            } else {
                uiState = .error(message: "Invalid response from server")
            }
            
            await verifyProgress()
        } catch {
            guard !isStaleOperation(operationId) else { return }
            
            let errorDescription = error.localizedDescription.lowercased()
            let isAlreadyCompleted = errorDescription.contains("already completed") || 
                                    errorDescription.contains("déjà complété")
            
            // Handle "already completed" error - treat as completion, not error
            if isAlreadyCompleted {
                uiState = .completed(message: "Onboarding already completed. Redirecting to home...")
            } else if let apiError = error as? APIError,
               case .httpError(let statusCode, let data) = apiError,
               statusCode == 400 {
                // Extract message from data if available
                if let data = data,
                   let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                   let message = json["message"] as? String {
                    let messageLower = message.lowercased()
                    if messageLower.contains("already completed") || messageLower.contains("déjà complété") {
                        uiState = .completed(message: "Onboarding already completed. Redirecting to home...")
                        return
                    }
                    uiState = .error(message: message)
                } else {
                    uiState = .error(message: error.localizedDescription)
                }
                return
            } else {
                uiState = .error(message: error.localizedDescription)
            }
        }
    }
    
    // MARK: - Submit Answer
    
    func submitAnswer(questionId: String, answer: AnswerValue) {
        let operationId = beginOperation()
        Task {
            do {
                guard let sessionId = currentSessionId else {
                    uiState = .error(message: "No active session")
                    return
                }
                
                uiState = .loading
                
                let request = AnswerRequest(
                    sessionId: sessionId,
                    questionId: questionId,
                    answer: answer
                )
                
                let response = try await OnboardingService.shared.submitAnswer(request)
                
                guard !isStaleOperation(operationId) else { return }
                
                if response.completed {
                    uiState = .completed(message: response.message ?? "Onboarding completed!")
                } else {
                    currentSessionId = response.sessionId
                    if let question = response.question, let progress = response.progress {
                        uiState = .questionLoaded(question: question, progress: progress)
                    } else {
                        uiState = .error(message: "Invalid response from server")
                    }
                }
                
                await verifyProgress()
            } catch {
                guard !isStaleOperation(operationId) else { return }
                uiState = .error(message: error.localizedDescription)
            }
        }
    }
    
    // MARK: - Skip Onboarding
    
    func skipOnboarding() {
        let operationId = beginOperation()
        Task {
            do {
                uiState = .loading
                let response = try await OnboardingService.shared.skipOnboarding()
                
                guard !isStaleOperation(operationId) else { return }
                
                uiState = .completed(message: response.message ?? "Onboarding skipped!")
            } catch {
                guard !isStaleOperation(operationId) else { return }
                uiState = .error(message: error.localizedDescription)
            }
        }
    }
    
    // MARK: - Reset Onboarding
    
    func resetOnboarding() async {
        let operationId = beginOperation()
        do {
            print("🔄 [OnboardingViewModel] Resetting onboarding...")
            uiState = .loading
            let response = try await OnboardingService.shared.resetOnboarding()
            
            guard !isStaleOperation(operationId) else {
                print("⚠️ [OnboardingViewModel] Reset operation is stale, ignoring")
                return
            }
            
            print("✅ [OnboardingViewModel] Reset successful - sessionId: \(response.sessionId), completed: \(response.completed), hasQuestion: \(response.question != nil)")
            currentSessionId = response.sessionId
            
            // After reset, if we get a question, show it. Otherwise, start a fresh onboarding
            if response.completed {
                print("ℹ️ [OnboardingViewModel] Reset returned completed status")
                uiState = .completed(message: response.message ?? "Onboarding completed!")
            } else if let question = response.question, let progress = response.progress {
                print("✅ [OnboardingViewModel] Reset returned question, showing it")
                uiState = .questionLoaded(question: question, progress: progress)
            } else {
                // Reset completed but no question returned - start fresh onboarding
                print("🔄 [OnboardingViewModel] Reset completed but no question, starting fresh onboarding...")
                await startOnboarding()
            }
            
            await verifyProgress()
        } catch {
            guard !isStaleOperation(operationId) else {
                print("⚠️ [OnboardingViewModel] Reset error but operation is stale, ignoring")
                return
            }
            print("❌ [OnboardingViewModel] Reset failed: \(error.localizedDescription)")
            // If reset fails, try to start anyway (maybe reset isn't needed)
            print("🔄 [OnboardingViewModel] Attempting to start onboarding after reset failure...")
            await startOnboarding()
        }
    }
    
    // MARK: - Verify Progress
    
    func verifyProgress() async {
        do {
            let status = try await OnboardingService.shared.getStatus()
            currentSessionId = status.sessionId ?? currentSessionId
            
            syncStatus = OnboardingSyncStatus(
                questionsAnswered: status.progress.questionsAnswered,
                canResume: status.canResume,
                lastUpdated: Date()
            )
            
            if status.onboardingCompleted {
                uiState = .completed(message: "Onboarding already completed. Redirecting to home...")
            }
        } catch {
            // Ignore sync errors - UI can retry on demand
            print("⚠️ [OnboardingViewModel] Failed to verify progress: \(error.localizedDescription)")
        }
    }
    
    // MARK: - Operation Management
    
    private func beginOperation() -> Int {
        let id = operationCounter.increment()
        latestOperationId = id
        return id
    }
    
    private func isStaleOperation(_ operationId: Int) -> Bool {
        return operationId != latestOperationId
    }
}

// MARK: - Atomic Integer Helper

class AtomicInteger {
    private var value: Int = 0
    private let queue = DispatchQueue(label: "com.wayfinder.atomic")
    
    func increment() -> Int {
        return queue.sync {
            value += 1
            return value
        }
    }
}
